common/timer.go

```diff
@@ -0,0 +1,206 @@
+package common
+
+import (
+	"math"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/common/types"
+)
+
+// TimerStore manages timers to efficiently support future timeouts. Timeouts
+// can be based on either block-height or block-timestamp. When a timeout occurs,
+// a designated callback function is called with the details (ctx and data).
+//
+// Once instantiated with NewFixationStore(), it offers the following methods:
+//    - WithCallbackByBlockHeight(callback): sets the callback for block-height timers
+//    - WithCallbackByBlockTime(callback): sets the callback for block-time timers
+//    - AddTimerByBlockHeight(ctx, block, data): add a timer to expire at block height
+//    - AddTimerByBlockTime(ctx, timestamp, data): add timer to expire at block timestamp
+//    - Tick(ctx): advance the timer to the ctx's block (height and timestamp)
+//
+// How does it work? The explanation below illustrates how the data is stored, assuming
+// the user is the module "package":
+//
+// 1. When instantiated, TimerStore gets a `prefix string` - used as a namespace to
+// separate between instances of TimerStore. For instance, module "packages" would
+// use its module name for prefix.
+//
+// 2. TimerStore keeps the timers with a special prefix; it uses the timeout value
+// (block height/timestamp) as the key, so that the standard iterator would yield them
+// in the desired chronological order. TimeSStore also keeps the next-timeout values
+// (of block height/timestamp) to efficiently determine if the iterator is needed.
+// For instance, module "packages" may have two (block height) timeouts with their data
+// set to "first" and "second" respectively:
+//
+//     prefix: packages_Timer_Next_       key: BlockHeight      value: 150
+//     prefix: packages_Timer_Next_       key: BLockTimer       value: MaxInt64
+//     prefix: packages_Timer_Value_          key: 150          data: "first"
+//     prefix: packages_Entry_Value_          key: 180          data: "second"
+//
+// 3. TimerStore tracks the next-timeout for both block-height/block-timestamp. On
+// every call to Tick(), it tests the current ctx's block height/timestamp against the
+// respective next-timeout:
+//
+// 4. If the next-timeout is reached/passed, then it will iterate through the timer
+// entries and invoke the (respective) callback for those entries; And finally it will
+// advance the (respective) next-timeout.
+
+// Thus, iterating on the prefix "packages_Entry_Index_" would yield all the package
```

Comment:
From here to the example, I'm not sure how this is related to the Timer explanation. Also, you have two "4."

Reply:
Sigh. Self copyright infringement :)
Will clean up.